### PR TITLE
feat: `ServerSentEventStream()`

### LINKED ISSUE #3693: proposal: replace `ServerSentEvent()` with `ServerSentEventStream()`
**Is your feature request related to a problem? Please describe.**

I am suggesting we replace the `http/server_sent_events.ts` implementation to be a simple `TransformStream` that performs the encoding of SSEs to strings.

**Describe the solution you'd like**

Here is an example I just threw together with @kt3k about how this should work:

```ts
const sse_stream = new SSEStream();

sse_stream.writable.getWriter().write({
  data: "Hello World",
  id: "",
  delay: 213,
});

sse_stream.readable.getReader().read();
// data: "Hello World"\r\n
// id: ""
// delay: 213\r\n\r\n

const stream = new ReadableStream({
  start(controller) {
    setInterval(() => {
      controller.enqueue({ data: "Hello World", id: "", delay: 213 });
    }, 1000);
  },
});

const body = stream.pipeThrough(new SSEStream()).pipeThrough(new TextDecoderStream())
new Response(body, { headers: { "Content-Type": "text/event-stream" } })
```

### PR DESCRIPTION
**Pull request by iuioiua**
You can test this via:
```ts
import {
  type ServerSentEventMessage,
  ServerSentEventStream,
} from "./http/server_sent_event_stream.ts";

Deno.serve({ port: 8000 }, (req) => {
  if (req.url.endsWith("/")) {
    const body = `
    <!DOCTYPE html>
    <html>
      <head>
        <meta charset=&quot;utf-8&quot; />
      </head>
      <body>
        <script>
          const source = new EventSource('/sse');
          source.onmessage = function(e) {
            document.body.innerHTML += e.data + '<br>';
          };
        </script>
      </body>
    </html>
    `;
    return new Response(body, {
      headers: { "content-type": "text/html; charset=utf-8" },
    });
  }

  if (req.url.endsWith("/sse")) {
    const stream = ReadableStream.from<ServerSentEventMessage>([{
      data: "hello there",
      id: "123",
    }]).pipeThrough(new ServerSentEventStream());
    return new Response(stream, {
      headers: {
        "content-type": "text/event-stream",
        "cache-control": "no-cache",
      },
    });
  }

  return new Response("not found", { status: 404 });
});
```

My main concern is that the name is too wordy. However, the reason I chose it is because it's clearest and doesn't conflict with the current `ServerSentEvent()` in `std/http`.

Closes #3693

### CODE REVIEW

File: http/server_sent_event_message_stream.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * Converts a server-sent message object into a string for the client.
+ *
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format}
+ */
+function stringify(message: ServerSentEventMessage) {
+  const lines = [];
+  if (message.comment) lines.push(`:${message.comment}`);
+  if (message.event) lines.push(`event:${message.event}`);
+  if (message.data) {
+    message.data.split(/\r?\n/).forEach((line) => lines.push(`data:${line}`));
```
Comment (lucacasonato) [inline http/server_sent_event_message_stream.ts:29]: I think just \r is also a valid new line delimiter. Please check the spec tho

File: http/server_sent_event_message_stream.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *   .pipeThrough(new TextEncoderStream());
+ * new Response(stream, {
+ *   headers: {
+ *     "content-type": "text/event-stream",
+ *     "cache-control": "no-cache",
+ *   },
+ * });
+ * ```
+ */
+export class ServerSentEventMessageStream
```
Comment (lucacasonato) [inline http/server_sent_event_message_stream.ts:60]:
```suggestion
export class ServerSentEventStream
```

**Review by lucacasonato — COMMENTED** [review]
Yay

File: http/server_sent_event_message_stream.ts (hunk 1 above)
Comment (iuioiua) [reply http/server_sent_event_message_stream.ts]: Yep, this line splits `messsage.data` by `\r\n`, `\n` and `\r`. In case there's any confusion, this is different to how newlines are represented in the SSE final message string, which is safe to just be a `\n`, according to the spec (see [here](https://html.spec.whatwg.org/multipage/server-sent-events.html#parsing-an-event-stream)). Either way, I've added to the test to check that `message.data` is split correctly for various delimiters.

File: http/server_sent_event_message_stream.ts (hunk 2 above)
Comment (iuioiua) [reply http/server_sent_event_message_stream.ts]: Done, but I've kept the `ServerSentEventMessage` interface name as-is.

File: http/server_sent_event_stream.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Converts a server-sent message object into a string for the client.
+ *
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format}
+ */
+function stringify(message: ServerSentEventMessage) {
+  const lines = [];
+  if (message.comment) lines.push(`:${message.comment}`);
+  if (message.event) lines.push(`event:${message.event}`);
```
Comment (lino-levan) [inline http/server_sent_event_stream.ts:28]: Are newlines allowed for these? Should we validate that these fields don't contain any newlines?

File: http/server_sent_event_stream_test.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    retry: 456,
+  }]).pipeThrough(new ServerSentEventStream());
+  let chunks = "";
+  for await (const chunk of stream) chunks += chunk;
+
+  assertEquals(
+    chunks,
+    `:a\nevent:b\ndata:c\ndata:d\ndata:e\ndata:f\nid:123\nretry:456\n\n`,
+  );
+});
```
Comment (lino-levan) [inline http/server_sent_event_stream_test.ts:23]: Would be nice to have a test that validated only some fields being specified.

**Review by lino-levan — APPROVED** [review]
Overall, looks great! Nice work Asher. Two nits.

File: http/server_sent_event_stream.ts (hunk 3 above)
Comment (iuioiua) [reply http/server_sent_event_stream.ts]: I've added a check that asserts that the single-line fields don't contain any newlines.

**Review by lino-levan — APPROVED** [review]
LGTM

**Comment by kt3k** [comment]
How about directly encoding into Uint8Array? Do we ever need string output of SSE?

**Comment by iuioiua** [comment]
Good idea. Done.

**Review by kt3k — APPROVED** [review]
LGTM!
